codec-http2/src/main/java/io/netty/handler/codec/http2/Http2MultiplexCodec.java

```diff
@@ -101,89 +79,132 @@ public Http2MultiplexCodec(boolean server, ChannelHandler streamHandler) {
+            }
+            fireChildReadAndRegister(childChannel, frame);
+        } else if (msg instanceof Http2GoAwayFrame) {
+            Http2GoAwayFrame goAwayFrame = (Http2GoAwayFrame) msg;
+            for (PrimitiveEntry<Http2StreamChannel> entry : childChannels.entries()) {
+                Http2StreamChannel childChannel = entry.value();
+                int streamId = entry.key();
+                if (streamId > goAwayFrame.lastStreamId() && isLocalStream(streamId)) {
+                    childChannel.pipeline().fireUserEventTriggered(goAwayFrame.retainedDuplicate());
+                }
+            }
+            goAwayFrame.release();
+        } else {
+            throw new UnsupportedMessageTypeException(msg);
+        }
+    }
+
+    void fireChildReadAndRegister(Http2StreamChannel childChannel, Http2StreamFrame frame) {
+        // Can't use childChannel.fireChannelRead() as it would fire independent of whether
+        // channel.read() had been called.
+        childChannel.fireChildRead(frame);
+        if (!childChannel.inStreamsToFireChildReadComplete) {
+            channelsToFireChildReadComplete.add(childChannel);
+            childChannel.inStreamsToFireChildReadComplete = true;
+        }
+    }
+
+    @Override
+    public void userEventTriggered(ChannelHandlerContext ctx, Object evt) throws Exception {
+        if (!(evt instanceof Http2StreamStateEvent)) {
+            ctx.fireUserEventTriggered(evt);
+            return;
+        }
+
+        int streamId = ((Http2StreamStateEvent) evt).streamId();
+        if (evt instanceof Http2StreamActiveEvent) {
+            onStreamActive(streamId);
+        } else if (evt instanceof Http2StreamClosedEvent) {
+            onStreamClosed(streamId);
+        } else {
+            throw new UnsupportedMessageTypeException(evt);
+        }
+    }
+
+    private void onStreamActive(int streamId) {
+        ChannelFuture future = createStreamChannel(ctx, streamId, streamHandler);
+        Http2StreamChannel childChannel = (Http2StreamChannel) future.channel();
+        childChannels.put(streamId, childChannel);
+    }
+
+    private void onStreamClosed(int streamId) {
+        final Http2StreamChannel childChannel = childChannels.remove(streamId);
+        if (childChannel != null) {
+            final EventLoop eventLoop = childChannel.eventLoop();
+            if (eventLoop.inEventLoop()) {
+                onStreamClosed0(childChannel);
+            } else {
+                eventLoop.execute(new Runnable() {
+                    @Override
+                    public void run() {
+                        onStreamClosed0(childChannel);
+                    }
+                });
+            }
+        }
+    }
+
+    private void onStreamClosed0(Http2StreamChannel childChannel) {
+        childChannel.onStreamClosedFired = true;
+        childChannel.fireChildRead(AbstractHttp2StreamChannel.CLOSE_MESSAGE);
+        childChannels.remove(childChannel.streamId);
```

Comment:
this is already done in `onStreamClosed` ... and also the eventloop in the child channel may be different and cause threading issues if we try to access the map here.